### PR TITLE
Machine allocation with dns and ntp

### PR DESCRIPTION
**Pull request by simcod**
This PR is related to [MEP-14](https://docs.metal-stack.io/dev/development/proposals/MEP14/README/).

Depends on:

- [x] https://github.com/metal-stack/metal-hammer/pull/139

### CODE REVIEW

File: test/inputs/goss.yaml [hunk 1]
```diff
@@ -54,6 +54,23 @@ file:
...
+    - server custom.1.ntp.org prefer iburst
+{{ else }}
+  "/etc/systemd/timesyncd.conf":
+    exists: true
+    contents:
+    - NTP=custom.1.ntp.org
+  "/etc/systemd/resolved.conf.d/dns.conf":
+    exists: true
+    contents:
+    - DNS=1.1.1.1
```
Comment (majst01) [inline test/inputs/goss.yaml:71]: Please add a test for /etc/resolv.conf as well

File: cmd/install.go [hunk 2]
```diff
@@ -171,9 +188,107 @@ func (i *installer) writeResolvConf() error {
 		`nameserver 8.8.8.8
 nameserver 8.8.4.4
 `)
+
+	if i.config != nil && i.config.DNSServers != nil {
```
Comment (majst01) [inline cmd/install.go:192]: This condition is somehow redundant to L73, please do it analog to L81-L86

File: cmd/install.go [hunk 3]
```diff
@@ -171,9 +188,107 @@ func (i *installer) writeResolvConf() error {
...
+	s.WriteString("\nLLMNR=no")
+
+	content := []byte(s.String())
+	return afero.WriteFile(i.fs, "/etc/systemd/resolved.conf.d/dns.conf", content, 0644)
+}
+
+func (i *installer) writeNTPConf() error {
+	var ntpConfigPath string
+	var content []byte
+	var s strings.Builder
```
Comment (majst01) [inline cmd/install.go:227]:
```suggestion
	var (
	   ntpConfigPath string
	   content []byte
	   s strings.Builder
	)

```

File: cmd/install.go [hunk 4]
```diff
@@ -171,9 +188,107 @@ func (i *installer) writeResolvConf() error {
...
+
+	switch i.config.Role {
+	case "firewall":
```
Comment (majst01) [inline cmd/install.go:230]:
```suggestion
	case  models.V1MachineAllocationRoleFirewall:
```

File: cmd/install.go [hunk 5]
```diff
@@ -171,9 +188,107 @@ func (i *installer) writeResolvConf() error {
...
+		for scanner.Scan() {
+			l := scanner.Text()
+			if !strings.HasPrefix(l, "pool") {
+				s.WriteString(l + "\n")
+			}
+		}
+		for _, ntp := range i.config.NTPServers {
+			s.WriteString("server " + ntp + " iburst\n")
+		}
+	case "machine":
```
Comment (majst01) [inline cmd/install.go:248]:
```suggestion
	case  models.V1MachineAllocationRoleMachine:
```

There are two more occurences of "firewall" and "machine", please replace them as well

**Review by majst01 — CHANGES_REQUESTED** [review]

File: test/inputs/goss.yaml [hunk 6]
```diff
@@ -54,6 +54,27 @@ file:
...
+{{ end }}
+  "/etc/resolv.conf":
+    exists: true
+    contents:
+    - nameserver 1.1.1.1
```
Comment (majst01) [inline test/inputs/goss.yaml:76]: Why is this only for !machine ?

Comment (simcod) [reply test/inputs/goss.yaml:76]: I thought this way I can apply the test, to check if the file exist, for all machine types.

**Review by majst01 — APPROVED** [review]

File: cmd/install.go [hunk 7]
```diff
@@ -155,23 +163,115 @@ func (i *installer) fileExists(filename string) bool {
...
-	return afero.WriteFile(i.fs, "/etc/resolv.conf", content, 0644)
+
+	if i.config.DNSServers != nil {
+		var s string
+		for _, dnsServer := range i.config.DNSServers {
+			s += "nameserver " + *dnsServer.IP + "\n"
+		}
+		content = []byte(s)
+
+		if i.oss != osAlmalinux {
```
Comment (Gerrit91) [inline cmd/install.go:190]: We actually overwrite the `/etc/resolved.conf` which causes systemd-resolved not to be used. So there is no need to write the `/etc/systemd/resolved.conf.d/dns.conf`.

**Review by Gerrit91 — APPROVED** [review]

**Pull request merged**